Oculars plugin: Introduce a 5-value pluginMode property

gzotti opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

### Description
<!--- Please include a summary of the change and which issue is fixed. -->
<!--- Please also include relevant motivation and context. -->
<!--- List any dependencies that are required for this change. -->

The mode switching and zooming code in the Oculars plugin is hard to follow, with a series of intermingled side effects.
This work here should consolidate the state switching. However, it seems the menu buttons in the top-right GUI panel still need boolean properties. I have moved those as "slave properties" with private getters/setters.

- better understandable mode switching with one centralized method.
- no more entangled side effects
- cleanup not finished yet. For now I have commented away the previous functions for easier comparison. 
- Any missing switching, storage and re-setting of main program settings should be done in the setPluginMode(). 

The other, even larger restructuring is the introduction of a dedicated Finder mode which replaces the hybrid isBinocular() from Ocular mode.  This adds a button to the menu bar. However, Binoculars and Finders are collected in the same class "Finder" given that their properties are specified in the same way. 

Fixes #1509, #356

### Screenshots (if appropriate):

### Type of change
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [x] This change requires a documentation update

### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

**Test Configuration**:
* Operating system: Windows 10
* Graphics Card: irrelevant

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the [code style](http://stellarium.org/doc/head/codingStyle.html) of this project.
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

gzotti committed:
Introduce a 4-value pluginMode property
- better understandable mode switching
- no more entangled side effects
- cleanup not finished yet

alex-w commented:
> ## 2021: Switching both possible modes off was the original logic. But what if we have an Ocular that is isBinocular()?
> https://github.com/Stellarium/stellarium/blob/b2601a66f250d1b6dcb7f5c0aaebe60af274f7b8/plugins/Oculars/src/Oculars.cpp#L850-L855
> 
> ###### This comment was generated by [todo](https://todo.jasonet.co) based on a `TODO` comment in [b2601a6](https://github.com/Stellarium/stellarium/commit/b2601a66f250d1b6dcb7f5c0aaebe60af274f7b8) in #1521. cc @Stellarium.

Binoculars and finderscopes are defined in eyepiece section by the historical reasons. According to  #356 we should introduce new tab in main GUI for finderscopes/binoculars and new button on on-screen panel (+ changes in ini files of course). I tried fix it some time ago, but without success, because fix was required seriously refactoring the code. :( By the fact you are started refactoring the parts of code, which need for the separation of binocular mode also.

gzotti commented:
Phew, this makes it a bit more ambitious. Then let's introduce a class Finder (maybe derivable from Ocular?) which can be used for binoculars as well (Of course we'll NOT make a funny OO screen mask for binos!) The point is its definition via characterization like "6x30" (mag x aperture) and also TRUE field of view, which may be given in degrees or "x meters in 1km distance" from terrestrial spotting scopes or binos. Do we have to support both? (It's just a arctangent operation. Users can be required to use a pocket calculator to convert their "x meters/km" to degrees, right?)  Actually, the "8x30" is pretty irrelevant to us, only TFOV counts. But in difference to Telrad, we can apply the estimate of limiting magnitudes like for Oculars, here aperture will count again. 

Shall this look like a Telrad plus screen mask plus crosshairs then? Do you have a useful icon? 

If you are in principle OK with this solution, I can remove a lot of comments and TODOs.

alex-w commented:
> Phew, this makes it a bit more ambitious. Then let's introduce a class Finder (maybe derivable from Ocular?) which can be used for binoculars as well (Of course we'll NOT make a funny OO screen mask for binos!) The point is its definition via characterization like "6x30" (mag x aperture) and also TRUE field of view, which may be given in degrees or "x meters in 1km distance" from terrestrial spotting scopes or binos. Do we have to support both? (It's just a arctangent operation. Users can be required to use a pocket calculator to convert their "x meters/km" to degrees, right?) Actually, the "8x30" is pretty irrelevant to us, only TFOV counts. But in difference to Telrad, we can apply the estimate of limiting magnitudes like for Oculars, here aperture will count again.

Yes, new class Finder is needed here and this class should contain the maths for visualisation the finders/binos

> Shall this look like a Telrad plus screen mask plus crosshairs then? Do you have a useful icon?

Yes and yes (see: https://github.com/Stellarium/stellarium/blob/master/plugins/Oculars/resources/bt_binocular_on.png)

> If you are in principle OK with this solution, I can remove a lot of comments and TODOs.

gzotti committed:
First cleanup

gzotti committed:
Typofix and declared slot

gzotti committed:
Add finder class.
- begin of restructuring
- runs, but Finder does nothing.
- need to move isBinocular() functionality from Ocular to Finder
- certainly many mode switching details unfinished

gzotti committed:
Fix a few simple errors

gzotti committed:
Improve Finder specs
- also simplify mappings in other instrument classes

gzotti committed:
Add even more trackable values...
switch handling still not complete yet.

gzotti committed:
Fix LGTM issues

gzotti committed:
Update comment

gzotti committed:
avoid a name conflict

gzotti commented:
Apparently superseded by #1633

gzotti commented:
Don't bother with the conflicts! This requires a rebase first.

gzotti committed:
Fixes after rebase 2023
